src/compliment/sources/cljs.cljc

```diff
@@ -0,0 +1,268 @@
+  [sym]
+  (some-> sym
+          (str "$macros")
+          symbol))
+
+(defn ns-macro-candidates
+  "Returns candidate data for all macros defined in ns."
+  [env ns]
+  (-> env
+      (ana/public-macros #?(:clj ns :cljs (ana/add-ns-macros ns)))
+      macro-candidates))
+
+(defn scoped-candidates
+  "Returns all candidates for the namespace of sym. Sym must be
+  namespace-qualified. Macro candidates are included if the namespace has its
+  macros required in context-ns."
+  [env sym context-ns]
+  (let [sym-ns (-> sym utils/as-sym utils/namespace-sym)
+        computed-ns (->ns env sym-ns context-ns)
+        macro-ns (->macro-ns env sym-ns context-ns)
+        sym-ns-as-string (str sym-ns)]
+    (mapcat #(prefix-candidates sym-ns-as-string %)
+            [(ns-public-var-candidates env computed-ns)
+             (when macro-ns
+               (ns-macro-candidates env macro-ns))])))
+
+(defn potential-candidates
+  "Returns all candidates for sym. If sym is namespace-qualified, the candidates
+  for that namespace will be returned (including macros if the namespace has its
+  macros required in context-ns). Otherwise, all non-namespace-qualified
+  candidates for context-ns will be returned."
+  [env sym context-ns]
+  (if (or (= (.indexOf ^String sym "/") -1) (.startsWith ^String sym ":"))
+    (unscoped-candidates env context-ns)
+    (scoped-candidates env sym context-ns)))
+
+(defn- distinct-candidates
+  "Filters candidates to have only one entry for each value of :candidate. If
+  multiple such entries do exist, the first occurrence is used."
+  [candidates]
+  (map first (vals (group-by :candidate candidates))))
+
+(defn- candidate-match?
+  [candidate prefix]
+
+  (.startsWith ^String (:candidate candidate) prefix))
+
+(defn- enrich-candidate
+  [candidate {:keys [context-ns extra-metadata cljs-env]}]
+  (if (seq extra-metadata)
+    (let [var-meta {} #_(i/info cljs-env (symbol (str (:ns candidate)) (:candidate candidate)) context-ns)]
+      (cond-> candidate
+        (and (:arglists extra-metadata) (:arglists var-meta))
+        (assoc :arglists (apply list (map pr-str (:arglists var-meta))))
+
+        (and (:doc extra-metadata) (:doc var-meta))
+        (assoc :doc (:doc var-meta))))
+    candidate))
+
+#?(:cljs
+   (defn- remove-candidate-macros
+     [candidate]
+     (if (= :namespace (:type candidate))
+       (update candidate :candidate (comp str u/remove-macros))
+       candidate)))
+
+(defn candidates
+  "Returns a sequence of candidate data for completions matching the given
+  prefix string and options in the ClojureScript compiler env.
+
+  - :cljs-env       - the ClojureScript compiler state;
+  - :context-ns     - (optional) namespace where completion is initiated as symbol;
+  - :extra-metadata - (todo) set of additional fields (:arglists, :doc) to add to the response maps."
+  [prefix {:keys [context-ns cljs-env] :as options-map}]   ;; WHAT TO DO HERE [^String prefix, ns context]
```

Comment:
Regarding the dynamic variable, how ugly is ugly? I would expect it to be bound outside, and here you can rebind it to a lexical variable and then pass it around. Seems OKish to me, anything I'm missing?

Reply:
No nothing missing, maybe it is ugly to me only 😁